정연희 - 자판기 온보딩 프로젝트

yeonhee7935 opened the pull request.

yeonhee7935 committed:
[feat] 기본 UI 생성

yeonhee7935 committed:
[feat] main.js: Vending Machine 초기화 로직 구현

- state 객체를 생성하여 애플리케이션 상태 관리(잔액, 로그, 상품)
- display, buttons, input, log 모듈 초기화 연결

yeonhee7935 committed:
[feat] display.js: 금액 표시창 업데이트 로직 추가

- 금액 표시창 초기화
- 금액 업데이트 함수 정의 및 state에 연결

yeonhee7935 committed:
[feat] buttons.js: 상품 버튼 동적 생성 및 클릭 이벤트 처리

- 상품 버튼 동적으로 생성
- 클릭 시 잔액 차감 및 금액 부족 시 처리 로직 추가

yeonhee7935 committed:
[feat] input.js: 금액 입력 및 투입 기능 구현

- 금액 입력 필드, 투입 버튼 이벤트 처리
- 입력값 검증 및 state 업데이트 처리
- 투입 후 필드 초기화

yeonhee7935 committed:
[feat] log.js: 동작 기록 및 로그 출력 기능 구현

- 로그 기록 및 UI 업데이트
- 최신 로그로 자동 스크롤 처리
- *로그 박스 높이 고정할 예정

yeonhee7935 committed:
[feat] 반환 버튼

yeonhee7935 committed:
[refactor] 마크업, 로그 수정

yeonhee7935 committed:
[feat] vite 설정 추가

yeonhee7935 committed:
[refactor] 폴더명 변경(vendingMachine -> src)

yeonhee7935 committed:
[refactor] 클래스화

yeonhee7935 committed:
[refactor] 레거시 코드 삭제

yeonhee7935 committed:
[refactor] data json파일로 빼기

yeonhee7935 committed:
[refactor] 클래스명 변경

yeonhee7935 committed:
[recycle] 잔액 업데이트 대기 중 버튼 클릭 시 타이머 취소 처리

- 금액 부족 시 부족한 금액을 띄우고 1.5초 후에 잔액을 표시함.
- 잔액으로 표시되기 전에 다른 제품 버튼을 누르면 해당 제품의 가격이 표시되나, 1.5초가 지나기 전에 잔액이 표시됨.
- (solution) 타이머의 id를 저장하고 있다가 버튼 클릭 시 이미 등록된 타이머가 있는 경우 타이머를 해제하는 로직 추가

yeonhee7935 committed:
[recycle] 제품 버튼을 이벤트 위임방식으로 변경

yeonhee7935 committed:
[refactor] Product class 삭제

yeonhee7935 committed:
[recycle] VendingMachine - init메소드 private으로 설정 / 클래스명 수정

yeonhee7935 committed:
[recycle] 금액투입 예외메시지 수정

taggon commented on src/index.js:
```diff
@@ -0,0 +1,15 @@
+
+async function loadProducts() {
+  const response = await fetch("/products.json");
+  return await response.json();
```
async 함수에서 리턴하는 await는 없어도 됩니다. :)

taggon commented on src/index.js:
```diff
@@ -0,0 +1,15 @@
+}
+
+document.addEventListener("DOMContentLoaded", async () => {
+  const products = await loadProducts();
```
이 코드에서 문제가 생길 여지는 없겠지만, 네트워크 함수를 사용할 때는 가능하면 에러 처리를 해주는 습관을 들여보면 어떨까요?

taggon commented on src/index.js:
```diff
@@ -0,0 +1,15 @@
+
+  const vendingMachine = new VendingMachine(products);
+
+  new VendingMachineUI(vendingMachine);
```
UI에 컨트롤러 역할을 하는 vendingMachine을 주입한 부분 👍

taggon commented on src/ui/Display.js:
```diff
@@ -0,0 +1,8 @@
+export class Display {
+  constructor() {
+    this.display = document.getElementById("display");
```
Display가 너무 특정 엘리먼트에 결합되어 있는 것으로 보입니다. 선택자, 아이디, 또는 객체를 전달받도록 수정하는 편이 나을 것 같아 보입니다.

taggon commented on src/ui/Display.js:
```diff
@@ -0,0 +1,8 @@
+    this.display = document.getElementById("display");
+  }
+  update(balance) {
+    this.display.textContent = balance.toLocaleString();
```
`toLocaleString()`으로 세 자리마다 콤마라는 요구사항을 준수하려면 사용자의 브라우저가 영어와 한국어 같이 세 자리 콤마를 지원하는 언어를 사용한다는 **가정**이 있습니다. 명시적으로 드러나지 않는 암묵적인 가정은 가능하면 없애는 편이 오류의 가능성도 줄이고 더 명확한 코드가 됩니다.

또한 `toLocaleString()` 보다는 `formatNumber`와 같은 함수로 래핑하는 편이 더 명확하게 보였을 것 같습니다.

taggon commented on src/ui/LogPanel.js:
```diff
@@ -0,0 +1,17 @@
+export class LogPanel {
+  constructor() {
+    this.logContainer = document.getElementById("log");
+    this.logs = [];
```
logs 배열의 역할은 무엇일까요?

taggon commented on src/ui/UIManager.js:
```diff
@@ -0,0 +1,120 @@
+
+  // 제품 버튼 생성
+  #createProductButton(product) {
+    const button = document.createElement("button");
```
직접 button 엘리먼트를 생성하고 클래스를 설정하는 대신 template 태그를 활용해보면 어떨까요?

taggon commented on src/ui/UIManager.js:
```diff
@@ -0,0 +1,120 @@
+          this.currentTimeoutId = null;
+        }
+
+        // 구매 실패 시, 가격을 표시하고 1.5초 뒤 잔액을 다시 표시
+        this.display.update(product.price);
+        this.currentTimeoutId = setTimeout(() => {
+          this.display.update(this.vendingMachine.getBalance());
+          this.currentTimeoutId = null; // 타이머 완료 후 초기화
+        }, 1500);
```
> 상품 버튼을 클릭했을 때 투입된 금액이 상품 가격보다 적으면 버튼을 _누르는 동안_ 금액 표시창에 상품 금액이 표시됩니다.

이 코드는 요구 사항과 맞지 않습니다.

taggon commented on src/ui/UIManager.js:
```diff
@@ -0,0 +1,120 @@
+      this.display.update(this.vendingMachine.getBalance());
+      this.logPanel.addLog(e.message);
+
+      if (errorCallback) {
+        errorCallback();
+      }
```
```suggestion
      errorCallback?.();
```
사소한 건데 이렇게 간단하게 사용할 수도 있습니다.

yeonhee7935 committed:
[refactor] logpanel, display에 객체 넘기도록 수정

yeonhee7935 committed:
[refactor] 제품조회 API 수정(loadProducts)

- return 시 await 구문 삭제
- 예외 발생 시 빈배열 반환

yeonhee7935 committed:
[refactor] errorCallback 호출 시 null chaining 적용

yeonhee7935 committed:
[refactor] 제품버튼 생성 시 template 태그 적용

yeonhee7935 committed:
[refactor] products 타입 변경( list -> dictionary )

yeonhee7935 committed:
[refactor] toLocaleString 수정

- 세 자리 마다 콤마(,) 찍기 -> locales 지정(지역별로 숫자 표현 방식이 다름)
- 해당 함수를 호출하는 부분을 별도의 함수로 추출

yeonhee7935 committed:
[refactor] 금액 부족 시 요구사항에 맞도록 수정

- (before) 1.5초 동안 상품 금액 표시
- (after) 버튼을 누르는 동안 상품 금액 표시

yeonhee7935 committed:
[refactor] UIManager -> VendingMachineUI

yeonhee7935 committed:
[refactor] 잔액 충분하면 클릭하자마자 구매되도록 수정

yeonhee7935 committed:
[refactor] 제품버튼 생성과 이벤트리스너 등록 로직 분리

yeonhee7935 committed:
[feat] 이벤트 환경 설정

yeonhee7935 committed:
[test] VendingMachine 모델 테스트코드 작성

yeonhee7935 committed:
[test] Display 테스트코드 작성

yeonhee7935 committed:
[test] LogPanel 테스트코드 작성

yeonhee7935 committed:
[refactor] 제품버튼 클래스명 변경(buttons -> products)

yeonhee7935 committed:
[test] index.js 테스트코드 작성

yeonhee7935 committed:
[test] VendingMachine 테스트 - 사전 상태 검증 추가

yeonhee7935 committed:
[refactor] 성공메시지 관리하는 클래스 추가

yeonhee7935 committed:
[refactor] formatString을 별도의 함수로 추출(테스트 시 사용하기 위해)

yeonhee7935 committed:
[refactor] listToDictionary 수정

yeonhee7935 committed:
[test] VendingMachine isNaN 테스트 조건 추가

yeonhee7935 committed:
[test] VendingMachine 테스트코드 작성

yeonhee7935 committed:
[test] 테스트코드 주석추가

yeonhee7935 committed:
[test] 제품버튼 클릭(마우스업, 다운) 테스트코드 작성

yeonhee7935 committed:
[test] 금액 투입 테스트코드 작성

yeonhee7935 committed:
[test] helper 함수 작성(insertMoney)
